ludwig/models/retrieval.py

```diff
@@ -0,0 +1,225 @@
+        raise NotImplementedError
+
+
+class RandomRetrieval(RetrievalModel):
+    """Random retrieval model.
+
+    Gets k random indices from the dataset regardless of the query.
+    """
+
+    def __init__(self, **kwargs):
+        self.index = None
+        self.index_data = None
+
+    def create_dataset_index(self, df: pd.DataFrame, backend: "Backend", columns_to_index: Optional[List[str]] = None):
+        self.index = np.array(range(len(df)))
+        self.index_data = df
+
+    def search(
+        self, df, backend: "Backend", k: int = 10, return_data: bool = False
+    ) -> Union[List[int], List[Dict[str, Any]]]:
+
+        results = []
+        for _ in tqdm(range(len(df))):
+            indices = np.random.choice(self.index, k, replace=False)
+
+            if return_data:
+                result = self.index_data.iloc[indices].to_dict(orient="records")
+            else:
+                result = indices
+            results.append(result)
+        return results
+
+    def save_index(self, name: str, cache_directory: str):
+        index_file_path = os.path.join(cache_directory, name + ".index")
+        # open file to prevent using the .npy extension
+        # https://numpy.org/doc/stable/reference/generated/numpy.save.html
+        with open(index_file_path, "wb") as f:
+            np.save(f, self.index)
+
+        index_data_file_path = os.path.join(cache_directory, name + "_data.csv")
+        self.index_data.to_csv(index_data_file_path, index=False)
+
+    def load_index(self, name: str, cache_directory: str):
+        index_file_path = os.path.join(cache_directory, name + ".index")
+        self.index = np.load(index_file_path)
+
+        index_data_file_path = os.path.join(cache_directory, name + "_data.csv")
+        self.index_data = pd.read_csv(index_data_file_path)
+
+
+class SemanticRetrieval(RetrievalModel):
+    """Semantic retrieval model.
+
+    Uses a sentence transformer model to encode the dataset and retrieve the top k most similar results to the query.
+    """
+
+    def __init__(self, model_name, **kwargs):
+        self.model_name = model_name
+        self.index: faiss.Index = None
+        self.index_data: pd.DataFrame = None
+        self.checksum = None
+
+        # best batch size computed during the encoding step
+        self.best_batch_size = None
+
+    def create_dataset_index(self, df: pd.DataFrame, backend: "Backend", columns_to_index: Optional[List[str]] = None):
+        if columns_to_index is None:
+            columns_to_index = df.columns
+        df_to_index = df[columns_to_index]
+
+        new_checksum = df_checksum(df_to_index)
+        if self.checksum == new_checksum:
+            # Reuse existing index
+            # TODO(travis): could use an LRU cache to support multiple datasets concurrently
+            return
+        self.checksum = new_checksum
+
+        rows = df_to_index.to_dict(orient="records")
+        row_strs = [json.dumps(r) for r in rows]
```

Comment:
nit - could potentially turn this into a helper function since search() also does the same thing

Reply:
Done